crates/sdl-encoder/src/directive_def.rs

```diff
@@ -0,0 +1,170 @@
+use crate::InputValue;
+use std::fmt::{self, Display};
+
+/// The `__Directive` type represents a Directive that a service supports.
+///
+/// *DirectiveDefinition*:
+///     Description<sub>opt</sub> **directive @** Name Arguments Definition<sub>opt</sub> **repeatable**<sub>opt</sub> **on** DirectiveLocations
+///
+/// Detailed documentation can be found in [GraphQL spec](https://spec.graphql.org/draft/#sec-Type-System.Directives).
+///
+/// ### Example
+/// ```rust
+/// use sdl_encoder::{Directive};
+/// use indoc::indoc;
+///
+/// let mut directive = Directive::new("infer".to_string());
+/// directive.description(Some("Infer field types\nfrom field values.".to_string()));
+/// directive.location("OBJECT".to_string());
+/// directive.location("FIELD_DEFINITION".to_string());
+/// directive.location("INPUT_FIELD_DEFINITION".to_string());
+///
+/// assert_eq!(
+///     directive.to_string(),
+///     r#""""
+/// Infer field types
+/// from field values.
+/// """
+/// directive @infer on OBJECT | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
+/// "#
+/// );
+/// ```
+#[derive(Debug)]
+pub struct Directive {
+    // Name must return a String.
+    name: String,
+    // Description may return a String or null.
+    description: Option<String>,
+    // Args returns a Vector of __InputValue representing the arguments this
+    // directive accepts.
+    args: Vec<InputValue>,
+    // Locations returns a List of __DirectiveLocation representing the valid
+    // locations this directive may be placed.
+    locations: Vec<String>,
```

Comment:
We may not need this right now, but it may make sense to use a `DirectiveLocation` enum here instead of `String`.

Reply:
Why would you want this as an enum? Is there other information about a directive we need to serialise?

Reply:
You'd want this as an enum if you're doing anything else than encoding it, like validating directive locations in a document. I also assume that would allow for a more compact representation and make it possible to store this as a set for easy membership checks for example.

Reply:
This module should not be doing anything other than encoding. Any sort of validation, checks or referencing should be done in other parts of the system given an AST.